### ISSUE TITLE
Build of "gecko" project timing out

### ISSUE THREAD

**Issue by indygreg** [rtfd/readthedocs.org#1874]
A few weeks ago, the build of https://gecko.readthedocs.org/ started timing out. Example output log is available at https://readthedocs.org/projects/gecko/builds/3585981/. I don't see any obvious errors. And I don't think the process stalled: these docs are just large and take a while to generate. I concede we're probably in the 99th percentile in terms of size :/

We try to use RTD as our canonical home for these Firefox development docs and people are starting to miss the up to date documentation.

Please advise on how to work around this timeout issue.

**Comment by indygreg** [comment]
ping?

**Comment by agjohnson** [comment]
Hrm, well none of the build commands are taking an excessive amount of time. My guess is there is either latency/network issues pulling the repository, or the repository is large enough that pulling it eats a good chunk of the allotted build time. Due to the amount of repo storage required, we have to cycle our repo cache very often, which likely explains why this is intermittent.

We're trying to find a fair balance here still. Putting in upper limits on builds has greatly improved service availability for the vast majority of users and reduced the operational burden on us. We do still need to find a fair solution to support projects, like this one, with larger demands though.

I've considered tracking monthly build time totals for projects and throttle build heavily after a threshold. We also lean towards a subscription that would guarantee more resources and maybe access to a separate build queue. Feedback on any of this is helpful, we're still trying to figure out how to balance free service and still make the service sustainable.

**Comment by martenson** [comment]
possibly related to https://github.com/rtfd/readthedocs.org/issues/1767

**Comment by agjohnson** [comment]
@indygreg recent builds look to have passed in an acceptable amount of time, seems this might have been a slow down after all?

I've bumped the project timeout for now, if this is still an issue the longer timeout may allow for enough time to build. Closing due to inactivity, but feel free to reopen if this is indeed still an issue, it might be a longer than expected delay pulling the repository.

**Comment by indygreg** [comment]
Yes, it looks to be fixed. Thank you!

The mozilla-central repository is massive. We're probably one of the largest repositories you interact with.

I added a feature to Mercurial (available in 3.6 and enabled by default in 3.7) called "clone bundles" that allows server operators to pre-generate repo bundles and have clients seed clones from static file servers (like S3 or a CDN). We have this deployed on hg.mozilla.org. From our testing at Mozilla, this makes clones significantly faster and more reliable. See https://www.mercurial-scm.org/wiki/Release3.7 and http://gregoryszorc.com/blog/2015/10/22/cloning-improvements-in-mercurial-3.6/ for more info. If you haven't upgraded to Mercurial 3.6 yet, you should do that to make `hg clone` and `hg pull` operations faster. (There are also some clone perf wins throughout versions 3.2+.) You may not notice much improvement on your average, small repo. But for large repos like mozilla-central the improvements could add up to dozens of seconds or even minutes of time saved.

**Comment by agjohnson** [comment]
Good point, we are using the Ubuntu 14.04 provided 2.8.2 version currently, I'm sure we could benefit from a more recent version. I've raised an issue to experiment with this in our server provisioning.

We've started investigating shallow cloning of git repos, I'm not sure if mercurial plans to have a similar feature. This could also reduce clone time and bandwidth significantly. It's not certain if the feature is usable for our case yet though.
